src/libs/elektra/keyhelpers.c

```diff
@@ -197,10 +197,13 @@ int keyNameIsSystem (const char * name)
  */
 int keyInit (Key * key)
 {
-	if (key->meta) ksClear (key->meta);
-	memset (key, 0, sizeof (struct _Key) - sizeof (KeySet *));
+	KeySet * ks = key->meta;
+	if (ks) ksClear (ks);
 
-	if (!key->meta) key->meta = ksNew (0, KS_END);
+	memset (key, 0, sizeof (struct _Key));
+	key->meta = ks;
+
+	if (!ks) key->meta = ksNew (0, KS_END);
```

Comment:
Can't we lazily alloc/initialize the meta keyset if needed? I think allocating the data structures for all keys is not needed. This additionally forces mmapstorage/cache to write all the empty keysets to disk, or allocate them while loading the cache. Maybe there are some other opinions?

Reply:
This is the discussion we had in #2983. See in particular https://github.com/ElektraInitiative/libelektra/issues/2983#issuecomment-534575198

The KeySet array is no longer allocated if called by `ksNew(0, KS_END)`. That is how we initialize `key->meta`, so that it isn't wasting so much memory.

We decided to always allocate `key->meta` because otherwise `keyMeta` would have been a much more complex function.

Reply:
What about:
```
KeySet * keyMeta (Key * key)
{
	if (!key) return 0;
	if (!key->meta) key->meta = ksNew (0, KS_END);

	return key->meta;
}
```
Is that too complex?

Reply:
This is also what I suggested https://github.com/ElektraInitiative/libelektra/issues/2983#issuecomment-534109621

@markus2330 wrote that with that you cannot check whether metadata exists at the moment. But thinking about it now, with the current solution there simply always exists metadata so you don't need that check and with your solution, too, just lazily. The main concern here was memory, but this solution is better memory-wise than to always create it, so it seems like an improvement to me.

Reply:
I'd prefer that approach, as it is transparent to the user anyway. That would allow `mmapstorage` to simply not store empty meta keysets. If the approach is eager then `mmapstorage` would need to persist this "small overhead per key" also to disk, which does not seem practical to me.

If there is a different reasoning which I am overlooking here, then of course my suggestion might not apply.

Reply:
Yes, I am also fine that we are lazy at two points in time:

1. key->meta might be empty if keyMeta was never called
2. key->meta->array might be empty if not meta keys were added

The implementation of @mpranj is fine. My concern is that the user should not get 0 if key->meta is null (as this complicates the use of this API, a second function or parameter is needed). But to create it lazily within keyMeta is fine. But I think this optimization (1.) will not have much effect as keyMeta will always be called at some point.

So instead of (1.) it might more sense to detect within mmap code that key->meta->array is not there, thus (in this case) the whole key->meta does not need to be persisted.

@mpranj what do you think about that?

Reply:
We can detect that inside mmapstorage, yes. But then we persist a nullpointer instead of the meta keyset, so we again need to lazily allocate it later. (Or eagerly allocate it inside mmap, which contradicts the mmap approach of trying to do as little as possible)

To me it sounds like a good solution to:
1. Filter empty meta keysets inside mmap.
2. Lazily allocate `key->meta` in the `keyMeta` function, s.t. mmap does not have to do it eagerly.

Reply:
Yes, sounds okay. In any case we can add both lazy allocations.